### ISSUE TITLE
`rustup doc` can't distinguish between keywords and primitives with the same name

### ISSUE THREAD

**Issue by cstyles** [rust-lang/rustup#2076]
**Problem**
First of all, I just want to say that I really like the recent addition of topics to the `rustup doc` feature! While playing around with it, I noticed that if I run `rustup doc fn`, it brings up the documentation for the keyword `fn`. However, there's also a primitive `fn` and I can't figure out any way to open that documentation page specifically. Am I missing something or is it just not possible?

This would also, in theory, affect any other naming collisions, not just with keywords and primitives (e.g., if some `Trait` and `Enum` had the same name and scope).

**Steps**

1. Run `rustup doc fn`
2. Observe that the `keyword.fn.html` page is opened in a browser.

**Possible Solution(s)**
I'm not super jazzed about this idea but we could add flags for `--keyword`, `--primitive`, `--macro`, etc. to allow the user to tell `rustup` exactly what they're looking for. But it would only ever be helpful in these rare situations so I don't know if it's worth A) the effort and B) making the UI more complex. It's also worth mentioning that on the page for the keyword `fn`, the first thing the viewer will see in the left-hand side panel is a list of primitives. So it probably wouldn't be too much of an inconvenience for them to just find `fn` there and click on it.

**Notes**

Output of `rustup --version`: `rustup 1.20.0 (2019-10-15)`
Output of `rustup show`:
```
Default host: x86_64-apple-darwin
rustup home:  /Users/collin.styles/.rustup

stable-x86_64-apple-darwin (default)
rustc 1.38.0 (625451e37 2019-09-23)
```

**Comment by kinnison** [comment]
I'm not sure I want to add more flags, but I also don't know how we'd distinguish them oherwise.  This is an interesting idea, but as you say it's not terribly hard for someone to navigate to the other concept.  I've re-labelled this help-wanted and enhancement rather than bug.

**Comment by cstyles** [comment]
The search function for the rust docs allows you to preface a search with something like `fn:` to restrict the search to just functions. This also works with `mod`, `struct`, `enum`, `trait`, `type`, `macro`, `const`. Perhaps we can do something like that here?

https://github.com/rust-lang/rust/blob/8e0007f829661e57d008d2e908c95f6e84b04b25/src/librustdoc/html/layout.rs#L134-L145

**Comment by kinnison** [comment]
Oh that is a very interesting approach yes.  This feature was originally written by @dorfsmay so perhaps they might have some input on how to do this?

**Comment by dorfsmay** [comment]
And I thought we really had been through all possible conflicts 😞!

So the idea would be `rustup doc primitive:fn` vs `rustup doc keyword:fn`? It shouldn't be too difficult to implement.

What should be the default though? What should `rustup doc fn` bring up? My own bias is keyword over primitive but I can value in the opposite. I'd really hate that there is no default and that we'd have to use the `type:` for all searches.

By the way, I just checked, `fn` is the only conflict between primitive and keywords at this point in time!

**Comment by kinnison** [comment]
If we go based on the sort order of the results of searching the stdlib docs for `fn` then the keyword should win over the primitive.  Do you think you can do a PR @dorfsmay ?

**Comment by dorfsmay** [comment]
The three types of docs in stdlib that are not module (which requires ::) are `keyword`, `macro` and `primitive`. The only conflict between those three currently is `fn`. IMO keyword before primitive make sense, but macro before keyword doesn't, should we future proof this a bit? Should we decide on the order and hard code it, rather than use alphabetical order?

**Comment by kinnison** [comment]
Yes, let's do that, keyword > primitive > macro I think

**Comment by dorfsmay** [comment]
Sounds good. I'll start working on it - not sure if needed or how to assign the issue to myself though.

**Comment by kinnison** [comment]
@dorfsmay I appreciate that life gets in the way sometimes, but I was wondering if you'd made a start on this?  I'm currently working my way through what I might want finished before the next release and this is on the list :D

**Comment by dorfsmay** [comment]
I was trying to replace the current test for topical doc by parametrized tests, one stand-alone test per data point, but it turned out to be a lot more complicated than I expected. I should be able to wrap this up within the next few days, I can re-visit parametrized tests later as a separate issue.

**Comment by kinnison** [comment]
That sounds amazing, thank you for your work on this. :beers:

**Comment by kinnison** [comment]
I believe this is resolved by #2119
